content/article/passing-by.md

```diff
@@ -0,0 +1,237 @@
+When assigning, `perl` often copies the value, see the following code:
+```perl
+my $var1 = "not modified";
+
+my $var2 = $var1; # Copy the content, not the variable itself
+$var1 = "modified";
+
+print "$var2\n";
+```
+This code will print the initial *non-modified* value:
+```
+not modified
+```
+
+It is because the assignment copies the content of the variable and does not *alias* the variable. 
+
+To better understand, you can check by printing the [reference](https://perldoc.perl.org/perlref) of the variable:
+```perl
+# Compare storage location
+print \$var1 . " vs " . \$var2 . "\n";
+```
+That gives 2 different storage locations:
+```perl
+SCALAR(0x55589c6378e0) vs SCALAR(0x55589c6378f8)
+                   ^^                        ^^  
+```
+So it clearly states that these are 2 different variables.
+
+But wait, there is also some cases where the assignment is actually *aliasing*, it is what happen in `foreach` loops:
+```perl
+my @array = ( "foo", "bar", "baz" );
+
+foreach my $var (@array) {
+    $var = "modified"; # Modify the initial @array
+}
+
+print "@array\n";
+```
+And the array is well modified:
+```
+modified modified modified
+```
+
+## A word about references
+A [reference](https://perldoc.perl.org/perlref) represents a single scalar value that holds *where-the-value-is-stored* informations.
+Having 2 different locations means 2 different variables.
+The string representation of a reference even gives details about what kind of value (scalar, array, hash...) is pointed. Like the `SCALAR(...)` just seen above:
+```
+SCALAR(0x...)
+```
+If you're only interested by the type pointed, you can get this information (e.g. for type checking) from the `ref` operator:
+```perl
+my @array = ();
+print ref(\@array) . "\n";
+```
+
+You can get a reference by using the [backslash operator](https://perldoc.perl.org/perlref#Backslash-Operator) and then dereference with sigils (e.g. `$`) or with `->`
+```perl
+my $str = "foo";
+
+# Get reference
+my $ref = \$str;
+
+print "$$ref\n";
+```
+Or 
+```perl
+my @array = ( "foo", "bar", "baz" );
+
+# Get reference
+my $ref = \@array;
+
+# Derefrence with ->
+print "$ref->[0]\n";
+```
+Both of them will print:
+```
+foo
+```
+
+You can even obtain a **reference of a reference** and later **dereference multiple times**. It enables the possibility to produce weird/dumb things like this:
+```perl
+my $str = "bazinga";
+
+# Get a ref on a ref on a ref on a...
+my $rrrrrrrrrref = \\\\\\\\\\$str;
+
+# De-de-de-...-de-dereference
+print "$$$$$$$$$$$rrrrrrrrrref\n";
+```
+
+But I repeat, don't do this.
+![](/images/passing-by/weird.png)
+
+## Modifications of variables into subs
+Inside subs, editing the variables won't usually ([\*](https://stackoverflow.com/a/5746000)) affect the arguments.
+```perl
+sub bycopy {
+    my $cp = shift;
+    $cp = "modified";
+}
+
+my $var = "not modified";
+bycopy($var);
+
+print "$var\n";
+```
+(for the curious, using `&` or a `prototype` won't change the game)
```

Comment:
I'd leave this out unless you were going to make a more important point about it.

Reply:
You confirm what I was thinking about this